src/gribtonc/gribtonc.1
." $Id: gribtonc.1,v 1.3 2003/02/21 21:37:46 rkambic Exp $
.TH GRIBTONC 1 "8 June 1995"
.SH NAME
gribtonc - GRIB to netCDF decoder
.SH SYNOPSIS
.HP
.ft B
gribtonc
.nh
\%[-l\ \fIlogpath\fP]
\%[-v]
\%[-t\ \fItimeout\fP]
\%[-e\ \fIerrpath\fP]
\%[-g\ \fIuserparams.tbl\fP]
\%[-q\ \fIqspecs\fP]
\%\fICDL_path\fP
\%\fInetCDF_path\fP
.hy
.ft
.SH DESCRIPTION
.LP
This program decodes GRIB products (embedded in WMO product envelopes) on
the NWS High-Resolution Data Service (HRS) feed, writing the decoded
products into a single netCDF output file, specified by
.IR netCDF_path .
In case the output netCDF file does not already exist, the
.I CDL_path
must be supplied to specify the structure for the desired netCDF file, which
.B gribtonc
creates. 
The default parameter translation is made according to a definition by UCAR,
but the user may supply specific translations in a separate file by the
-g tabfile option. The user-supplied translations will override the default
ones. 
GRIB products are converted to the units specified for the appropriate
netCDF variable, if this differs from the units used for the GRIB parameter.
A set of CDL files for all the model outputs appearing on the HRS data
stream is provided with
.B gribtonc
for use in decoding HRS GRIB data.
.LP
.SH FORMAT OF GRIB TO NETCDF TRANSLATION TABLE
.LP
The format of the translation table is as follows:
Each line contains
.RS +4
.nf 
GRIB parameter number
netCDF parameter name (cannot start with digit, [A-Za-z0-9_] only )
netCDF long description
netCDF units
An optional comment
.fi
.RE

These are all separated by semicolon (";"). To comment a whole line, 
put any letter or special character (i.e. not a digit) in front of it.
Trailing spaces are ignored before each semicolon, but not after.
The GRIB number may have spaces in front.
Parameters found in the separate file take presidence over the 
default parameter table, 
those not listed in the separate file are unchanged.
All parameters will be named with a level suffix in addition to the name
assigned in the separate file.

.SH EXAMPLE
The command
.RS +4
.nf
gribtocdl -v -g my.tab sst.grib > sst.cdl
.fi
.RE

will decode the products in sst.grib and generate the corresponding CDL
code into file sst.cdl. 
The file my.tab contains a specific conversion table to be used when
naming a parameter from GRIB.  
Later the GRIB file may be converted to netCDF
by using gribtonc:
.RS +4
.nf
gribtonc -v -g my.tab sst.cdl sst.nc < sst.grib
.fi
.RE

The file my.tab looks like this:
.RS +4
.nf
Grib;netCDF;long name               ;units;comments
 31 ;CI    ;Sea-ice cover           ;none ;(0-1)
 34 ;SSTK  ;Sea surface temperature ;K    ;
151 ;MSL   ;Mean sea-level pressure ;Pa   ;
167 ;T2M   ;2 metre temperature     ;K    ;
.fi
.RE
.LP
.LP
.B gribtonc
can run as a subprocess of the LDM \fBpqact\fP(1) program, receiving a stream
of GRIB products from standard input and decoding them as they arrive.
.B gribtonc
can also be used as a standalone decoder, decoding products on its standard
input into netCDF form, as specified by
.I netCDF_file
GRIB products that do not conform to the specified file structure
(e.g. they are defined on a different grid or contain a grid parameter not
included in the specified netCDF file) are skipped.
.LP
To use
.B gribtonc
from the LDM \fBpqact\fP(1) program requires adding entries  
to the \fBpqact\fP configuration file
that invoke 
.B gribtonc
on appropriate HRS products.  The PIPE action should be used in pattern-action
entries.  The form of such an entry is:
.HP
  HRS	\fIpattern\fR		PIPE	\fIgribtonc-executable\fR
				\fIgribtonc-options\fR
				\fIcdl-file\fR	\fInetcdf-file\fR
.LP
where \fIpattern\fR is a regular expression pattern matching product
identifiers of products that are to be decoded and written to the output
file \fInetcdf-file\fR.  
.LP
Any directories that do not exist in \fInetcdf-file\fR
will be created as needed, so date components, for example, may be part of
the output file path, specified by using matched parts of the regular
expression \fIpattern\fR.  In case the output \fInetcdf-file\fR does not
already exist, it will be created from the CDL specification in the file
specified by \fIcdl-file\fR.
.LP
The path of the executable for 
.B gribtonc
must be specified as the first
argument to the PIPE command, and should be either an absolute path name or
in a directory in the executable search path of \fBpqact\fP.
.LP
An example \fBpqact.conf\fP entry for decoding GRIB products from the ETA model as
they appear on the HRS data stream is:
.HP
# ETA model on Lambert conformal grid to yyyymmddhh_eta.nc
HRS	^Y.Q... KWBC (..)(..).*/mETA
	PIPE	/usr/local/ldm/decoders/gribtonc
	etc/eta.cdl
	data/GRIB/(\1:yyyy)(\1:mm)\1\2_eta.nc
.LP
Since 
.B gribtonc
reads GRIB products from its standard input, it may be invoked independently
of \fBpqact\fP(1) by using I/O redirection.
.SH OPTIONS 
.TP
.BI "-l " logpath
Log file pathname.
The program uses Unidata's \fBulog\fP(3) package to write error and log
messages.
The default is to write to the \fBsyslogd\fP(8) daemon.  If this option is
specified, however, then messages will be written to the file
\fIlogpath\fP unless that pathname is "-", in which case all messages will be
written to standard error.
.TP
.B -v
Verbose logging.
Informative messages (level \fBLOG_INFO\fP) are logged, including a line for
each GRIB product decoded.
By default, only messages of severity \fBLOG_NOTICE\fP and greater are
logged.
.TP
.BI \-t " timeout"
Timeout in seconds (default is 600).  If there has been no input available
on stdin for the duration of the specified timeout, the program exits
gracefully.
.TP
.BI \-e " errpath"
If a GRIB product cannot be decoded, perhaps because it is truncated or
otherwise corrupted, it is appended to this file.
.TP
.BI \-g " tabfile"
Use GRIB-netcdf translation defined in this file. The format of this file
is described below.
.TP
.BI \-q " qspecs"
Specification for how to expand so-called quasi-regular or thinned grids
into rectangular grids.  This argument is a
comma-delimited text string that specifies the method of interpolation used
(currently \fBlin\fP for linear or \fBcub\fP for cubic), and the latitude
and longitude
increments in degrees for the resulting grid (e.g. \fBdlat=2.5,dlon=5.0\fP).
.TP
.B CDL_path
The path name of a file to be used as a template for the netCDF file, if the
netCDF file specified as output does not exist.  If the netCDF output file
does exist, this argument is ignored.
A path name relative to the directory
specified by the LDM_ETCDIR environment variable may be used, but you must
make sure this environment variable is defined in 
.B gribtonc
's environment.
.TP
.B netCDF_path
The path name of the netCDF output file.
This may be specified as an absolute path name or (more
typically) relative to the directory from which 
.B gribtonc
is invoked.
Directories that do not exist in the path name of the output file
will be created as needed.

.SH EXAMPLE
The command
.RS +4
.nf
gribtonc -v -l- -q "lin,dlat=2.5,dlon=5.0" avn-x.cdl avn-x.nc < avn-x.wmo
.fi
.RE
will decode the products in avn-x.wmo into the netCDF file
avn-x.nc, creating the latter file if it does not already exist from
the ASCII CDL template file avn-x.cdl.  If the data contain quasi-regular
grids, they will be expanded to full regular grids on a 2.5 by 5.0 degree
grid before being written into the output file.
.SH "SEE ALSO"
gribdump(1), netcdf(3), pqact(1), udunits(3), ulog(3)
.SH DIAGNOSTICS
Error messages and log messages are written to the log file.



src/RELEASE-NOTES
$Id: RELEASE-NOTES,v 1.37 2004/12/21 20:55:26 rkambic Exp $

Release	Date		Notes
_________________	______________________________________________________
5.0.2 2013-05-07	Corrected gfs_37-44.cdl (grids are 1.25x1.25 degree);
                        added ukmet_21-24.cdl (like GFS AVN, these are 1.25x1.25 degree
                        grids); added '-n' flag to gribtonc (says to not expand
                        the record dimension past what is defined in the CDL/existing
                        netCDF file); updated 'port' subdirectory with latest code
                        (so that the package will build on CentOS 6.4 x86_64)
5.0.1 2012-06-13	Remade ruc2_236.cdl because of transfer of the RUC2 production
			from FSL to RAP 
5.0.0 2007-10-10	The perl Decoders have been moved to their own package
			so the process of deprecating this package can take place.
			This package is being deprecated because of it's 
			limitation of not being able to decode Grib2, compile
			problems, and static parameter table.
4.1.7 2007-09-27	metar2nc decoders removes reports with illegal cntl chars
			changed dist from Z to gz format
4.1.6 2007-08-20	fixed typo in eta_211.cdl
4.1.5 2007-02-09	added a -c option to metar2nc* so it can process archive
                        data from the command line.
4.1.4 2006-10-04	fixed bug in ua2nc, putting 23hr data into wrong file.
4.1.3 2006-09-14	fixed bug in ua2nc, putting 23hr data into wrong file.
4.1.2 2006-09-06	added more levels
4.1.1 2006-08-23	minor bug fixes, recs.c added import, metar2nc.new1
			SPD and GUST are now floats when converted from knots
			to m/s. etc
4.0.0 2005-12-19	update the METAR table sfmetar_sa.tbl, created new
			metar.cdl.new1, new table lsfstns.tbl for synoptic 
			decoder, and new table for upperair decoder snstns.tbl
			updated other cdls. the new table names cause mods to
			syn2nc and ua2nc.
4.0.0 2005-10-05	changed perl from /usr/local/bin/perl to /usr/bin/perl
			pqact entries don't need yyyymm argument anymore.
			new naming convention, ie surface_METAR_yyyymmdd_hhmm.nc
			added -h option to write hourly files for the all perl
			decoders, default is daily file. updated sfmetar.tbl, 
			snworld.tbl and systns_upc.tbl. fixed bug in syn2nc 
			that returned wind DIR greater than 900, removed java 
			Grib decoder jars, it's has it own release package now
			called grib-decoder-java_2.2.10.tar.gz. gfs_211.cdl,
			eta_211.cdl, and ruc2_236.cdl were updated to add
			new parmeters. managed log files to only keeps 3 logs
			added global attributes to metar.cdl.new1
			changed testdata.wmo to testdata.wmo.bak, etc

3.1.1 2005-05-24	changed units of Dx and Dy to be meters instead of km 
			in all cdls. changed cdl version to be 1.0
3.1.0 2005-05-24	fixed the perl decoders so unicode characters don't put
			decoder into infinite loop. This fix is needed for 
			perl version 5.8.0 and higher. fixed buoy2nc and syn2nc
			where min was set to 60 and timegm() was called. 
			Also added the java grib.jar to the release.

3.0.9 2005-03-01	fixed the perl decoder cdls to be udunits compliant.  
3.0.8 2005-02-04	fixed metar2nc and metar2nc.new1 dew point temperature
			incase the coding changes from /M00 to /M0
3.0.7 2004-12-21	added level 245, changed files: levels.h gribtocdl.c 
			levels.c nc.c
3.0.6 2004-10-21	some parameters started with a number, thats illegal
			in netcdf. also correct units for oztot (ozone total)
3.0.5 2004-10-18	update cdls and new version of metar.new1, metar.cdl.new

3.0.4 2004-07-22	rewrote perl decoders to be more robust especially
			when checking the report time for accuracy. Changed
			test to read in more stations from the sfmetar_sa.tbl
			station file. Added metar2nc.new1 and metar.cdl.new1 as
			a new perl decoder, new netCDF format. New script
			makeCDL1.pl makes a cdl file metar.cdl.new1. In recs.c
			upped RECS_INIT_SIZE to 64. Updated cdls
			In nc.c added stronger test to parmname, levels
			mods to test.pl for real time data, mods to
			src/*/testdata.wmo for real time testing. Fixed 
			return value for time range indicator in gribtocdl

3.0.3 2004-03-12	added Writer scripts into all the perl decoders. These
			scripts bust-up bulletins into individual reports
			added missing level info to gribtocdl.c program
			added Makefile.in to zone MANIFEST
			Fixed century bug, files recs.c units.c gribdump

3.0.2 2004-02-01	change #define MAX_GRIB_SIZE 200000
			added forecasttime variable to gribtocdl
			rename all cdls to reflect new model names
			updated INSTALL and CUSTOMIZE files
			in cdl dir, updated Makefile.in and README
			mods to metar2nc, made matches more robust

3.0.1 2003-04-25	added CF standard_name, forecasttime, and enhanced
			Conventions attribute to acknowledge CF conventions
			to all model cdls. Modified recs.c to write out
			forecasttime.

3.0.0 2003-03-25	added dimension valtime_offset for the valtime_offset
			variable. Enhanced cdl to have the data for the 
			valtime_offset var.  This permits the data to be stored
			in monotonic increasing order.  Added parameter_number
			and level_number attributes to variables.

2.4.4 2003-02-25        added vars gust_sfc to ruc2.cdl and vars Tmin_fhg and
			Tmax_fhg to avn-x.cdl
			mods to gdes.c gribtocdl.[1c] inetutil.c ulog.h so 
			code compiles on AIX boxes.

2.4.3 2003-02-21        modifications so user parms functionality works
			correctly in the code, new functionality to 
			gribdump, also updated the man pages. files:
			gribtocdl.c, params.c, gribdump.c, gribtonc.1 and
			gribdump.1
			
2.4.2 2003-02-19        Added the PNO indicator for Metar decoder
			In gdes.c, changed i_dim from lat to lon
			fixed avn-1.0x1.0.cdl also

2.4.1 2003-01-31        Fixed some error for the avn-1.0x1.0 Grid.  Also fixed
			errors for the Gausian lat/lon grid, plus other fixes
			mods to gribtocld.c gdes.c nc.c 

2.4.0 2002-11-04        Merge in code made by Arild Burud and Terje Reite for

src/gribtonc/gribtonc.c
/*
 *	Copyright 1993, University Corporation for Atmospheric Research.
 */
/* $Id: gribtonc.c,v 1.14 2002/12/05 20:00:04 rkambic Exp $ */

/* 
 * decodes GRIB products into netCDF files
 * modified May 2002 by Arild Burud/NoSerC, allow user to specify 
 * grib parameter conversion by "-g gribtablefile"
 */
int testNoserc;

#include <stdio.h>
#include <signal.h>
#include <errno.h>
#include <unistd.h>
#include <limits.h>		/* _POSIX_PATH_MAX */
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include "ulog.h"
#include "mkdirs_open.h"

#include "nc.h"
#include "centers.h"
#include "models.h"
#include "params.h"
#include "levels.h"
#include "timeunits.h"
#include "grib1.h"
#include "product_data.h"
#include "quasi.h"
#include "units.h"
#include "user_param.h"

#ifdef NO_ATEXIT
#include <atexit.h>
#endif

#ifdef __STDC__
static void cleanup(void);
static void signal_handler(int sig);
static void set_sigactions(void);
static void usage(char* av0);
static void print_floats(float* ff, int nn, int pp);
static int is_ixg(int id);
static int do_nc(FILE * ep, int to, quas*, char *cdlpath, char *ncpath);
static product_data* grib_decode(prod*, quas*);
int main(int ac, char** av);
#endif

int nogrow=0; /* grow/no grow unlimited dimension beyond what is defined in the input CDL */

unsigned long num_wmo_messages; /* for statistics on exit */
unsigned long num_gribs_decoded;
unsigned long num_gribs_written;
unsigned long num_degribbed;

/*
 * Timeout in seconds.  If no input is received for this interval, process
 * closes output file and exits.
 */
#define DEFAULT_TIMEOUT  600


/*
 * Called at exit.
 * This callback routine registered by atexit().
 */
static void
cleanup()
{
    uinfo("Exiting") ;
    uinfo("%lu WMO msgs, %lu GRIBs decoded, %lu written ",
	  num_wmo_messages, num_gribs_decoded, num_gribs_written);
    nccleanup();		/* close open netCDF files, if any */
    (void) closeulog();
}


/*
 * Called upon receipt of signals.
 * This callback routine registered in set_sigactions() .
 */
static void
signal_handler(sig)
     int sig ;
{
#ifdef SVR3SIGNALS
    /* 
     * Some systems reset handler to SIG_DFL upon entry to handler.
     * In that case, we reregister our handler.
     */
    (void) signal(sig, signal_handler) ;
#endif
    switch(sig) {
      case SIGHUP :
	udebug("SIGHUP") ;
	return ;
      case SIGINT :
	unotice("Interrupt") ;
	exit(0) ;
      case SIGTERM :
	udebug("SIGTERM") ;
	exit(0) ;
      case SIGUSR1 :
	udebug("SIGUSR1") ;
	return ;
      case SIGUSR2 :
	if (toggleulogpri(LOG_INFO))
	  unotice("Going verbose") ;
	else
	  unotice("Going silent") ;
	return ;
      case SIGPIPE :
	unotice("SIGPIPE") ;
	exit(0) ;
    }
    udebug("signal_handler: unhandled signal: %d", sig) ;
}


static void
set_sigactions()
{
#ifndef NO_POSIXSIGNALS
    struct sigaction sigact ;
    
    sigact.sa_handler = signal_handler;
    sigemptyset(&sigact.sa_mask) ;
    sigact.sa_flags = 0 ;
#ifdef SA_RESTART	/* SVR4, 4.3+ BSD */
    /* usually, restart system calls */
    sigact.sa_flags |= SA_RESTART ;
#endif
    
    (void) sigaction(SIGHUP, &sigact, NULL) ;
    (void) sigaction(SIGINT, &sigact, NULL) ;
    (void) sigaction(SIGTERM, &sigact, NULL) ;
    (void) sigaction(SIGUSR1, &sigact, NULL) ;
    (void) sigaction(SIGUSR2, &sigact, NULL) ;
    (void) sigaction(SIGPIPE, &sigact, NULL) ;
#else
    (void) signal(SIGHUP, signal_handler) ;
    (void) signal(SIGINT, signal_handler) ;
    (void) signal(SIGTERM, signal_handler) ;
    (void) signal(SIGUSR1, signal_handler) ;
    (void) signal(SIGUSR2, signal_handler) ;
    (void) signal(SIGPIPE, signal_handler) ;
#endif
}


static void
usage(av0)
     char *av0 ; /*  id string */
{
    (void)fprintf(stderr,
		  "Usage: %s [options] [files ...]\t\nOptions:\n", av0);
    (void)fprintf(stderr,
		  "\t-v           Verbose, report decoding steps\n") ;
    (void)fprintf(stderr,
		  "\t-l logfile   Send log info to file (default uses syslogd)\n") ;
    (void)fprintf(stderr,
		  "\t-g tabfile   Use grib-netcdf translation defined in this file\n") ;
    (void)fprintf(stderr,
		  "\t-t timeout   If no input, exit after \"timeout\" seconds (default %d)\n",
		  DEFAULT_TIMEOUT) ;
    (void)fprintf(stderr,
		  "\t-e errfile   Append bad GRIB products to this file\n") ;
    (void)fprintf(stderr,
		  "\t-n           Do not grow unlimited dimension beyond CDL define\n") ;
    (void)fprintf(stderr,
		  "\t-q meth      Specs for expanding quasi-regular grids,\n\t\t\te.g. -q \"lin,dlat=2.5,dlon=5.0\"\n") ;
    (void)fprintf(stderr,
		  "\t[ CDL_file ] CDL template, when netCDF output file does not exist\n") ;
    (void)fprintf(stderr,
		  "\tnetCDF_file  netCDF output file\n") ;
    exit(1);
}


/*
 * Parse raw product bytes into product_data structure.  Returns 0 if
 * failed.  User should call free_product_data() on result when done with
 * it.  Also expands quasi-regular grids to full rectangular grids if qmeth
 * is non-null.
 */
static product_data *
grib_decode(prodp, quasp)
     prod *prodp ;		/* input raw product bytes */
     quas *quasp ;		/* if non-null, method used to expand
				   quasi-regular "grids" */
{
    grib1 *gp = new_grib1(prodp); /* overlay raw bits on raw grib1 structure */
    product_data *pdp;

    if (gp == 0)
	return 0;

    pdp = new_product_data(gp);	/* compute cooked product structure, with GDS
				   (manufactured, if necessary) and bytemap */
    
    free_grib1(gp);

    if (!pdp)
	return 0;

    if (pdp->gd->quasi && quasp) {
	int ret = expand_quasi(quasp, pdp) ; /* Changes *pdp */
	if (!ret) {
	    uerror("%s: can't expand quasi-regular grid", pdp->header);
	    return 0;
	}
    }
    return pdp;
}


static int
do_nc (ep, timeout, quasp, cdlname, ncname)
    FILE *ep;			/* if non-null, where to append bad GRIBs */
    int timeout;		/* exit if no data in this many seconds */
    quas *quasp;		/* if non-null, specification for how
				   quasi-regular "grids" are to be expanded */
    char *cdlname;		/* Pathname of CDL template file to be used to
				   create netCDF file, if it doesn't exist */
    char *ncname;		/* Pathname of netCDF output file */
{
    struct prod the_prod;	/* raw bits of GRIB message, length, id */
    struct product_data *gribp;	/* decoded GRIB product structure */
    FILE *fp = stdin;		/* input */
    ncfile *ncp;
    int ncid;
    
    ncid = cdl_netcdf(cdlname, ncname);	/* get netCDF file handle */
    if (ncid == -1) {
	uerror("can't create or size 0 output netCDF file %s, exiting", 
	ncname);
	exit(-1);
    }

    setncid(ncid);	/* store ncid so can be closed if interrupt */

    num_wmo_messages = 0;
    num_gribs_decoded = 0;

    if (init_udunits() != 0) {
	uerror("can't initialize udunits library, exiting");
	exit(-1);
    }
    ncp = new_ncfile(ncname);
    if (!ncp) {
	uerror("can't create output netCDF file %s, exiting", ncname);
	exit(-1);
    }

    while(1) {			/* usual exit is timeout in get_prod() */
	int bytes = get_prod(fp, timeout, &the_prod);
	if (bytes == 0)
	  continue;
	else
	  num_wmo_messages++;
	
	/* decode it into a grib1 product structure */
	gribp = grib_decode(&the_prod, quasp);
	if (gribp == 0) {
	    if (ep) {		/* doesn't write WMO header, just product */
		if (fwrite(the_prod.bytes, the_prod.len, 1, ep) == 0) {
		    serror("writing bad GRIB to error file");
		}
	    }
	    continue;
	}
	num_gribs_decoded++;

	/* Write grib info to netCDF file */
	if (nc_write(gribp, ncp) == 0) {/* OK */
	    num_gribs_written++;
	}
	free_product_data(gribp);
    }
}


/*
 * Reads GRIB data from standard input.
 * GRIB data may be contained in WMO envelope or not.
 * Decodes GRIB data to in-memory structure.
 * Writes decoded data that conforms to one specified netCDF output file.
 * May be used as an in-line LDM decoder, invoked by pqact(1).
 */
int
main(ac,av)
     int ac ;
     char *av[] ;
{
    char *logfname = 0 ;	/* log file name, default uses syslogd */
    char *ofile = 0 ;		/* output netCDF file name */
    char *cdlfile = 0 ;		/* CDL template file name */
    FILE *ep = 0;		/* file handle for bad GRIBS output, when
				   -e badfname used */
    int timeo = DEFAULT_TIMEOUT ; /* timeout */
    quas *quasp = 0;		/* default, don't expand quasi-regular grids */

    /*
     * register exit handler
     */
    if(atexit(cleanup) != 0)
      {
	  serror("atexit") ;
	  exit(1) ;
      }
    
    /*
     * set up signal handlers
     */
    set_sigactions() ;
    
    {
	extern int nogrow;
	extern int optind;
	extern int opterr;
	extern char *optarg;
	int ch;
	int logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_NOTICE)) ;
	int errflg = 0;
	
	opterr = 1;
	
	while ((ch = getopt(ac, av, "vxl:t:e:q:g:n")) != EOF) {
	    switch (ch) {
	    case 'v':
		logmask |= LOG_MASK(LOG_INFO) ;
		break;
	    case 'x':
		logmask |= LOG_MASK(LOG_DEBUG) ;
		break;
	    case 'l':
		logfname = optarg ;
		break;
	    case 'g':
		user_makeparamtable(optarg) ;
		break;
	    case 'n':  /* <<<<< UPC mod 20130505 - grow/no grow unlimited dimension */
		nogrow = 1;
		break;
	    case 't':
		timeo = atoi(optarg) ;
		if(timeo < 1) {
		    fprintf(stderr, "%s: invalid timeout %s",
			    av[0], optarg) ;
		    errflg++;
		}
		break;
	    case 'e':
		ep = fopen(optarg, "w");
		if(!ep) {
		    serror("can't open %s", optarg);
		    errflg++;
		}
		break;
	    case 'q':
		quasp = qmeth_parse(optarg);
		if(!quasp) {
		    fprintf(stderr,
			    "%s: invalid quasi-regular expansion method %s",
			    av[0], optarg) ;
		    errflg++;
		}
		break;
	    case '?':
		errflg++;
		break;
	    }
	}
	
	switch (ac - optind) {	/* number of args left */
	  case 1:		/* must be existing netCDF file */
	    ofile = av[optind] ;
	    break;
	  case 2:
	    cdlfile = av[optind] ;
	    ofile = av[optind+1] ;
	    break;
	  default:
	    errflg++;
	}

	if(errflg)
	  usage(av[0]);	
	(void) setulogmask(logmask) ;
    }
        
    /*
     * initialize logger
     */
    (void) openulog(ubasename(av[0]),
		    (LOG_CONS|LOG_PID), LOG_LOCAL0, logfname) ;
    uinfo("Starting Up") ;

    do_nc(ep, timeo, quasp, cdlfile, ofile);
    
    return 0;
}

